test/client/spec/helpers/constrain-tick-labels.spec.js

```diff
@@ -0,0 +1,433 @@
+
+const mockMeasureText = function * (measurements) {
+  const mapper = (measurement) => (tick, index) => {
+    return {
+      text: tick,
+      measurement: measurement[index],
+      index
+    };
+  };
+  for (const measurement of measurements) {
+    const tickValues = yield;
+    yield tickValues.map(mapper(measurement));
+  }
+};
+
+const mockMeasureTicks = (generator) => {
+  return (props) => {
+    const result = generator.next(props.tickValues).value;
+    generator.next();
+    return result;
+  };
+};
+
+const defaultProps = {
+  orientation: "top",
+  style: {
+    axis: {
+      stroke: "black"
+    },
+    grid: {
+      strokeWidth: 2
+    },
+    ticks: {},
+    tickLabels: {
+      fontSize: 18
+    }
+  },
+  tickValues: [
+    "Mets\nNY",
+    "Giants\nSF",
+    "Yankees\nNY",
+    "Nationals\nDC",
+    "Mariners\nSEA"
+  ],
+  height: 300,
+  padding: 50,
+  scale: "linear",
+  standalone: true,
+  tickCount: 5,
+  width: 450
+};
+
+const defaultLayoutProps = {
+  style: {
+    parent: {
+      height: 300,
+      width: 450
+    },
+    axis: {
+      stroke: "black",
+      fill: "none",
+      strokeWidth: 2,
+      strokeLinecap: "round"
+    },
+    axisLabel: {
+      stroke: "transparent",
+      fill: "#756f6a",
+      fontSize: 16,
+      fontFamily: "Helvetica"
+    },
+    grid: {
+      fill: "none",
+      strokeLinecap: "round",
+      strokeWidth: 2
+    },
+    ticks: {
+      fill: "none",
+      padding: 5,
+      strokeWidth: 2,
+      strokeLinecap: "round",
+      size: 4
+    },
+    tickLabels: {
+      stroke: "transparent",
+      fill: "#756f6a",
+      fontFamily: "Helvetica",
+      fontSize: 18,
+      padding: 5
+    }
+  },
+  padding: {
+    top: 50,
+    bottom: 50,
+    left: 50,
+    right: 50
+  },
+  orientation: "top",
+  isVertical: false,
+  labelPadding: 0,
+  offset: {
+    x: 50,
+    y: 50
+  }
+};
+
+const defaultTickProps = {
+  ticks: [1, 2, 3, 4, 5],
+  stringTicks: true
+};
+
+describe("constrainTickLabels", () => {
+  describe("with horizontal axis", () => {
+    it("uses ellipses to prevent label collision", () => {
+      const normalWidth = defaultProps.width / defaultProps.tickValues.length;
+      const height = defaultLayoutProps.offset.y;
+      const constants = Array(4).fill({
+        width: `${normalWidth}px`, height: `${height}px`
+      });
+      const measureTextGenerator = mockMeasureText([
+        [...constants, {width: `${normalWidth + 20}px`, height: `${height}px`}],
+        [...constants, {width: `${normalWidth + 10}px`, height: `${height}px`}],
+        [...constants, {width: `${normalWidth}px`, height: `${height}px`}],
+        [...constants, {width: `${normalWidth}px`, height: `${height}px`}]
+      ]);
+      measureTextGenerator.next();
+      const ticks = constrainTickLabels({
+        props: defaultProps,
+        layoutProps: defaultLayoutProps,
+        tickProps: defaultTickProps,
+        measureFunc: mockMeasureTicks(measureTextGenerator)
```

Comment:
Generators that take parameters in next() are super weird and require these sorts of gymnastics. Section 4.1 (here)[http://www.2ality.com/2015/03/es6-generators.html] explains some of the strangeness involved!